pkg/daemon/daemon.go

```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
-// handleNodeUpdate is the handler for informer callbacks detecting node
-// changes. If an update is requested by the controller, we assume that
-// that means something changed and apply the desired config no matter what.
+// runOnce pulls the MachineConfig from either a file or remote URL
+// executes one run, and exits.
+// TODO: Revisit the Run/process methods and refactor and unify them with runOnce
+func (dn *Daemon) runOnce() error {
+	var machineConfig *mcfgv1.MachineConfig
+	var oldConfig mcfgv1.MachineConfig
+	var err error
+
+	if strings.HasPrefix("http://", dn.onceFrom) || strings.HasPrefix("https://", dn.onceFrom) {
+		// NOTE: This case expects a cluster to exists already.
+		// If we sense a remote URL has been provided then request MC content
+		// from a remote URL and parse it.
+		glog.V(2).Infof("Getting machine config content from %s", dn.onceFrom)
+		machineConfig, err = dn.getMachineConfigFromURL(dn.onceFrom)
+		if err != nil {
+			return err
+		}
+		needUpdate, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return err
+		} else if needUpdate == false {
+			glog.V(2).Infof("No update needed")
+			return nil
+		}
+		// At this point we have verified we need to update
+		if err := dn.executeUpdateFromClusterWithMachineConfig(machineConfig); err != nil {
+			glog.Warningf("Unable to update: %s", err)
+			return err
+		}
+		return nil
+
+	} else if ValidPath(dn.onceFrom) {
+		// NOTE: This case expects that the cluster is NOT CREATED YET.
+		// If we sense a local file has been provided parse it.
+		oldConfig = mcfgv1.MachineConfig{}
+		absoluteOnceFrom, err := filepath.Abs(filepath.Clean(dn.onceFrom))
+		if err != nil {
+			return err
+		}
+		machineConfig, err = dn.getMachineConfigFromFile(absoluteOnceFrom)
+		if err != nil {
+			return err
+		}
+		// Execute update without hitting the cluster
+		return dn.update(&oldConfig, machineConfig)
+	}
+	// Otherwise return an error as the input format is unsupported
+	return fmt.Errorf("%s is not a path nor url; can not run once", dn.onceFrom)
+}
+
+// handleNodeUpdate is the gatekeeper handler for informer callbacks detecting
+// node changes. If an update is requested by the controller, we assume that
+// that means something changed and pass over to execution methods no matter what.
 // Also note that we only care about node updates, not creation or deletion.
 func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 	node := cur.(*corev1.Node)
 
 	// First check if the node that was updated is this daemon's node
-	if (node.Name != dn.name) {
-		// The node that was changed was not ours
-		return
+	if node.Name == dn.name {
+		// Pass to the shared update prep method
+		_, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return
+		}
+		if err = dn.executeUpdateFromCluster(); err != nil {
+			glog.V(2).Infof("Unable to execute update: %s", err)
+			return
```

Comment:
same as above

Reply:
:+1:

Reply:
Re-checking the code `executeUpdateFromCluster()` will mark the node degraded and reboot upon error. I'll add some comments though.